### PR TITLE
Make the output of `ninja -t inputs` deterministic

### PR DESCRIPTION
**Pull request by digit-google**
This sorts the output of `ninja -t inputs` to make it
deterministic and remove duplicates.

### CODE REVIEW

**Comment by jhasse** [comment]
LGTM

@naarcini what do you think?

**Review by naarcini — APPROVED** [review]
Generally LGTM % a couple of suggestions

File: src/ninja.cc [hunk 1]
```diff
@@ -766,9 +770,30 @@ int NinjaMain::ToolInputs(const Options* options, int argc, char* argv[]) {
...
   }
 
-  set<Edge*> seen;
+  std::set<Edge*> seen;
+  std::vector<std::string> result;
   for (vector<Node*>::iterator in = nodes.begin(); in != nodes.end(); ++in)
-    PrintInputs((*in)->in_edge(), &seen);
+    CollectInputs((*in)->in_edge(), &seen, &result);
+
+  // Make output deterministic by sorting then removing duplicates.
```
[the remark this answers is not in the document]
Comment (digit-google) [reply src/ninja.cc:778]: I think I'm confused, I though std::unique was not part of C++03 !? if it is I would gladly use it (the code is based on the implementation example on https://en.cppreference.com/w/cpp/algorithm/unique)

[the remark this answers is not in the document]
Comment (naarcini) [reply src/ninja.cc:778]: I found it difficult to find good documentation about what is or isn't in C++03, but I believe std::unique is present. I'm basing this partly on the cppreference page for the algorithm library, which annotates which version of C++ particular functions were added (no annotation for `unique`):
https://en.cppreference.com/w/cpp/algorithm

And partly by clicking through the gcc github repo, which includes an implementation for std::unique in version 2.95 (last edited 1998):
https://github.com/gcc-mirror/gcc/blob/releases/gcc-2.95/libstdc%2B%2B/stl/stl_algo.h

**Comment by digit-google** [comment]
I have added a unit-test for Edge::CollectInputs. It seems there are no good ways to test tools though. Should we want to add these, should we go with regression tests (e.g. scripts that invoke Ninja with pre-determined build plans, and verify behaviour), or refactor ninja.cc to make tools unit-testable? Probably  not in this PR though :-)

**Comment by jhasse** [comment]
Some very basic regression tests are in `misc/output_test.py`, e.g.: https://github.com/ninja-build/ninja/blob/cd967ab21eef51819473b0cea22135fead76ffc2/misc/output_test.py#L117

Making tools unit-testable is probably a good idea though.

**Comment by digit-google** [comment]
Thanks, I could add a regression test to `output_test.py` for the new tool.

I'm still concerned about the fact that we do not include implicit inputs in the the tool's output. The fact that they are not listed in the command line itself doesn't make them less important. And I don't see a use case where you would want only the explicit inputs anyway. should we add an "explicit-inputs" tool for these otherwise?

**Comment by naarcini** [comment]
@digit-google In my specific use case, we only needed the explicit dependencies. At the time, I was trying to improve upon a project in Chromium that audits certain metadata in code (Network Traffic Annotation Auditor). In that case, all I needed was specifically the code files that are used to build a particular target. Since Chromium's tooling does a pretty good job of explicitly listing all dependencies, I found that "explicit" was enough to fulfill this requirement.

I'm not opposed to adding an additional flag or changing the default behavior to include more information. That'd still work fine for my use case and I suspect it'd work fine for most other peoples' use cases too. If I get a vote, I vote for using an additional flag instead of multiple tool names :)

**Comment by digit-google** [comment]
Latest update adds `--type=TYPE` option to the tool, as in `-t inputs --type=TYPE <targets>`. The default is to print explicit + implicit inputs, but this can be changed with `--type=explicit` and `--type=all`. I think that should address all use cases now :)

**Comment by digit-google** [comment]
Friendly ping, anything else to get this merged? I would really prefer if the existing behavior was not part of the next 1.11 release :-/

**Comment by jhasse** [comment]
The `--type` flag needs to be split into its own commit. But before you do that: Is the flag really necessary? I would like to keep the inputs tool simple for the moment.

Also:

> (i.e. this allows us to skip any stale outputs from a previous build, since cleandead will not work in certain cases).

Sounds like a bug of the cleandead tool?

**Comment by digit-google** [comment]
Well, I have a use case where I need all implicit + explicit dependencies, and @naarcini only wants the explicit ones, so a `--type` flag seems the better option here.

This is not really a bug in `cleandead`, it is more a Ninja design issue: the `.ninja_deps` file will record all entries generated by depfiles, and sometimes, some of them will be stale (because a command changed and will no longer use that file to the build and the corresponding depfile). Ninja has no way to detect these cases before re-running the build, so `cleandead` will always keep the stale files in the build directory. This has been causing a variety of breakages in our incremental builds (e.g. https://fxbug.dev/83397 or more recently https://fxbug.dev/85465.

This tool would be one of the different ways we want to work around this problem.

I'll update the PR to make the `--type  flag its own commit.

File: src/graph_test.cc [hunk 2]
```diff
@@ -215,6 +215,53 @@ TEST_F(GraphTest, RootNodes) {
...
+TEST_F(GraphTest, CollectInputs) {
+  ASSERT_NO_FATAL_FAILURE(AssertParse(
+      &state_,
+      "build out$ 1: cat in1 in2 in$ with$ space | implicit || order_only\n"));
+
+  std::vector<std::string> inputs;
+  Edge* edge = GetNode("out 1")->in_edge();
+
+  inputs.clear();
+  edge->CollectInputs(false, InputsType::Default, &inputs);
```
Comment (jhasse) [inline src/graph_test.cc:227]: This needs to be part of the second commit.

File: src/ninja.cc [hunk 3]
```diff
@@ -211,29 +213,31 @@ struct Tool {
...
+      "this system]\n"
+      "  -k N     keep going until N jobs fail (0 means infinity) [default=1]\n"
+      "  -l N     do not start new jobs if the load average is greater than N\n"
+      "  -n       dry run (don't run commands but act like they succeeded)\n"
+      "\n"
+      "  -d MODE  enable debugging (use '-d list' to list modes)\n"
+      "  -t TOOL  run a subtool (use '-t list' to list subtools)\n"
+      "    terminates toplevel options; further flags are passed to the tool\n"
+      "  -w FLAG  adjust warnings (use '-w list' to list warnings)\n",
+      kNinjaVersion, config.parallelism);
```
Comment (jhasse) [inline src/ninja.cc:241]: Why this change?

File: src/ninja.cc [hunk 4]
```diff
@@ -744,31 +748,66 @@ int NinjaMain::ToolCommands(const Options* options, int argc, char* argv[]) {
   return 0;
 }
 
-void PrintInputs(Edge* edge, set<Edge*>* seen) {
+void CollectInputs(Edge* edge, set<Edge*>* seen,
```
Comment (jhasse) [inline src/ninja.cc]: while you're at it, can you add `std::` here?

File: src/ninja.cc [hunk 5]
```diff
@@ -31,13 +31,15 @@
 #include <unistd.h>
 #endif
 
+#include <algorithm>
```
Comment (jhasse) [inline src/ninja.cc:34]: can you move this include to the top?

File: src/ninja.cc [hunk 6]
```diff
@@ -31,13 +31,15 @@
...
+
 #include "browse.h"
 #include "build.h"
 #include "build_log.h"
-#include "deps_log.h"
 #include "clean.h"
 #include "debug_flags.h"
 #include "depfile_parser.h"
+#include "deps_log.h"
```
Comment (jhasse) [inline src/ninja.cc:42]: any reason for this reordering?

File: src/ninja.cc [hunk 7]
```diff
@@ -771,19 +772,36 @@ int NinjaMain::ToolInputs(const Options* options, int argc, char* argv[]) {
...
+      }
+      if (!inputs_type.ParseArg(optarg))
+        Fatal("Invalid --type value, see '-t inputs --type=list'");
+      break;
     case 'h':
     default:
       printf(
           "Usage '-t inputs [options] [targets]\n"
           "\n"
-          "List all inputs used for a set of targets. Note that this ignores\n"
```
Comment (jhasse) [inline src/ninja.cc]: The "List all inputs used for a set of targets." shouldn't be removed, right?

**Review by jhasse — CHANGES_REQUESTED** [review]
> and @naarcini only wants the explicit ones

If I understood him correctly, he wouldn't mind the implicit ones, but doesn't need them?

The extra `InputsType` struct is quite a lot of code and I only merged the inputs tool because it was such a small change.

**Comment by digit-google** [comment]
I pushed a new version with the changes you requested 10 days ago, but forgot to submit my review comments, which I just did. Sorry. Let me know if there is something else to do to get this submitted. Thanks.

File: src/ninja.cc (hunk 6 above)
Comment (jhasse) [reply src/ninja.cc:42]: I see. Can you still move it back? Best to avoid unnecessary noise in the diff.

File: src/ninja.cc [hunk 8]
```diff
@@ -211,6 +213,7 @@ struct Tool {
 
 /// Print usage information.
 void Usage(const BuildConfig& config) {
+  // clang-format off
```
Comment (jhasse) [inline src/ninja.cc:217]: Please don't add those, clang-format is not enforced in Ninja's code base.

**Review by jhasse — COMMENTED** [review]
I still think that the tool shouldn't include the `--type` option and always print all inputs, at least for 1.11.0.

**Comment by digit-google** [comment]
Done, let me know what you think :)

**Comment by digit-google** [comment]
I have refactored the two commits so that the first one would print all inputs in a deterministic way (and update the documentation), while the second one adds the `--type=TYPE` option. I'm just waiting for the checks here to complete, and when everything works, I'll move the second commit to a different PR so that the first one can be part of the 1.11 release as you requested.

**Comment by digit-google** [comment]
Done, the current PR only contains the fix to the output. The `--type` option was moved to my Github fork as https://github.com/digit-google/ninja/tree/inputs-type which I will rebase on top of upstream when this one is merged.
Let me know if this requires further changes. This now prints all inputs by default.

**Review by jhasse — APPROVED** [review]

**Comment by jhasse** [comment]
Thanks!

**Pull request merged**